kong/runloop/plugins_iterator.lua

```diff
@@ -454,61 +463,79 @@ function PluginsIterator.new(version)
+      local cfg = get_plugin_config(plugin, name, plugin.ws_id)
+      if cfg then
         has_plugins = true
-      end
 
-      plugins[name] = true
+        if HANDLERS_WITH_CONFIGURE[name] then
+          configurable[name] = configurable[name] or {}
+          configurable[name][#configurable[name] + 1] = cfg
+        end
 
-      -- Retrieve route_id, service_id, and consumer_id from the plugin object, if they exist
-      local route_id = plugin.route and plugin.route.id
-      local service_id = plugin.service and plugin.service.id
-      local consumer_id = plugin.consumer and plugin.consumer.id
+        local data = ws[plugin.ws_id]
+        if not data then
+          data = new_ws_data()
+          ws[plugin.ws_id] = data
+        end
+        local plugins = data.plugins
+        local combos = data.combos
 
-      -- Get the plugin configuration for the specified workspace (ws_id)
-      local cfg = get_plugin_config(plugin, name, plugin.ws_id)
-      -- Determine if the plugin configuration is global (i.e., not tied to any route, service, consumer or group)
-      local is_global = not route_id and not service_id and not consumer_id and plugin.ws_id == kong.default_workspace
-      if is_global then
-        -- Store the global configuration for the plugin in the 'globals' table
-        globals[name] = cfg
-      end
+        plugins[name] = true
+
+        -- Retrieve route_id, service_id, and consumer_id from the plugin object, if they exist
+        local route_id = plugin.route and plugin.route.id
+        local service_id = plugin.service and plugin.service.id
+        local consumer_id = plugin.consumer and plugin.consumer.id
+
+        -- Determine if the plugin configuration is global (i.e., not tied to any route, service, consumer or group)
+        if not (route_id or service_id or consumer_id) and plugin.ws_id == kong.default_workspace then
+          -- Store the global configuration for the plugin in the 'globals' table
+          globals[name] = cfg
+        end
 
-      if cfg then
         -- Initialize an empty table for the plugin in the 'combos' table if it doesn't already exist
         combos[name] = combos[name] or {}
 
         -- Build a compound key using the route_id, service_id, and consumer_id
         local compound_key = build_compound_key(route_id, service_id, consumer_id)
 
-        -- Store the plugin configuration in the 'combos' table using the compound key
+       -- Store the plugin configuration in the 'combos' table using the compound key
         combos[name][compound_key] = cfg
       end
     end
+
     counter = counter + 1
   end
 
-  for _, plugin in ipairs(loaded_plugins) do
-    local name = plugin.name
-    for _, data in pairs(ws) do
-      local plugins = data.plugins
-      if plugins[name] then
-        local n = plugins[0] + 1
-        plugins[n] = plugin
-        plugins[0] = n
-        plugins[name] = nil
+  if has_plugins then
```

Comment:
What about reversing the logic to reduce multi-level nesting?

``` lua
if not has_plugins then
... abort
end

for _, plugin...

...
end
```

Reply:
It is hardish.